implement ext-foreign-toplevel-list-v1 by augmenting our existing implementation of foreign toplevels

sodiboo opened the pull request:
- https://github.com/sodiboo/niri/commit/d2d51bcf0ceb6d47fe5b8985268cf0902a2e921d but brought back from the archives.
- duplicate of #1450 but taking a different approach

i originally didn't submit this as a PR, because i didn't realize there were any client apps that depended on *just toplevel list*: i thought this protocol was useless on its own. but apparently cosmic needs it?

Besides technical reasons why one approach is better than the other, this implementation has one clear concrete difference over #1450: the `identifier` of a foreign toplevel is stable and can be used to manipulate a toplevel with `niri msg`, they are the same identifier. It's not actually clear that this is a *desirable* trait: do we want to allow clients to parse this identifier? In #1450, the foreign toplevel identifier is random bytes.

sodiboo committed:
implement ext-foreign-toplevel-list-v1

YaLTeR reviewed (commented):
I think I also prefer this approach to the other one. Though, I must note that ext-foreign-toplevel-list won't necessarily grow toplevel management functionality (rather, separate protocols may be added). See:
- https://github.com/quickshell-mirror/shell-protocols/pull/1
- https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/156
- https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/196

That is to say, the naming here could be somewhat adjusted, in particular the `WLR_VERSION`-`EXT_VERSION`-like pairs, which imply it's ~the same protocol (like, say, wlr/ext data control), but it isn't. (I don't want the names to get too unwieldy, but I also don't want to rename everything in the future when one more of these protocols arrives.)

The code looks ok at a glance, but I didn't read through it in-depth, which I will still need to do before merging.

> Besides technical reasons why one approach is better than the other, this implementation has one clear concrete difference over #1450: the `identifier` of a foreign toplevel is stable and can be used to manipulate a toplevel with `niri msg`, they are the same identifier. It's not actually clear that this is a _desirable_ trait: do we want to allow clients to parse this identifier? In #1450, the foreign toplevel identifier is random bytes.

I don't have an opinion on this; while it's useful, there's nothing that says other compositors should also match their IPC identifiers with this, so it ends up being niri-specific either way.

sodiboo commented on src/window/mapped.rs:
```diff
@@ -183,6 +183,27 @@ impl MappedId {
+    ///
+    /// This brings the total length to 27 characters for u64::MAX.
+    pub fn to_protocol_string(self) -> String {
+        format!("mapped-{}", self.0)
```
Realistically, i think this `mapped-` prefix adds nothing: that is only because it derives from a struct called `MappedId`, but that's kind of an internal naming? We could get away with just the numeric ID as a string here: if we need other namespaces for any reason, adding a prefix *is* fully backwards compatible. But when would we ever need another namespace?

sodiboo replied:
I've gone ahead and changed it to just be the ID string, without a prefix.

Like i said, we can add prefixes in the future (or a "don't parse" me random bytes implementation also cannot ever overlap, if it is fixed 32 length).

sodiboo commented on src/protocols/foreign_toplevel.rs:
```diff
@@ -1,31 +1,37 @@
     display: DisplayHandle,
-    instances: Vec<ZwlrForeignToplevelManagerV1>,
+    ext_instances: Vec<ExtForeignToplevelListV1>,
+    wlr_instances: Vec<ZwlrForeignToplevelManagerV1>,
```
We use a hashset for tracking handles for each toplevel; this should probably also be a hashset. Makes `Stop` implementation more efficient, as it's a hashset removal instead of `Vec::retain`

sodiboo committed:
clean up foreign toplevel destruction: don't `retain` to remove one item

sodiboo committed:
document why we have duplicate destruction logic in foreign toplevel

messy to have two of the same comment but like how else do i make that
information readily available to both

sodiboo commented:
> I don't have an opinion on this; while it's useful, there's nothing that says other compositors should also match their IPC identifiers with this, so it ends up being niri-specific either way.

Well, yes. Of course. Even if this *was* the common practice and encouraged in protocol documentation, it would *still* be niri-specific, because for this property to be useful, a client needs to know how to make use of the niri IPC interface.

Although, reading through the documentation, i notice now one minor concern with my current implementation:

> It is recommended that a compositor includes an opaque generation value in identifiers

This is mainly related to compositor handoff, but can even apply to us. Multiple-invocation clients with "one-shot" connections, such as many CLI apps, are allowed to *store* an identifier and persist it.

If it's a "dumb" implementation like i have, exposing an internal identifier, and the session is handed off to another process (e.g. session restore: we are still in the *same session*), then there's no guarantee that the new compositor allocates toplevels in the same order, and existing identifiers may now refer to a different toplevel.

Of course, niri doesn't support any such session restore. But it can also be an issue if such an app simply thinks it is the same session, but niri has restarted in the time since. Then, the identifiers may be reused with no notice; of course they can! it's a new session!

The "opaque generation value" can be based on e.g. the startup time of the compositor.

# Other compositors

- COSMIC ([usage](https://github.com/pop-os/cosmic-comp/blob/a74c90f14fa32e53f7abdaaadf3774b3da86cb19/src/wayland/protocols/toplevel_info.rs#L337)), as well as #1450 ([usage](https://github.com/bbb651/niri/blob/b9c08ab00c517965c2d073917f4962f82201f9ed/src/handlers/compositor.rs#L169)) both use the default provided by Smithay ([implementation](https://github.com/Smithay/smithay/blob/ace98eb20dfffc2f24ea5681bfc6112b5ccc8956/src/wayland/foreign_toplevel_list/mod.rs#L325-L335)) which does random bytes. Smithay also exposes a way to provide an identifier, but nobody seems to be using it.

- wlroots doesn't seem to support specifying a stable identifier. They have a generic [`token`](https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/c14aa1d0b82b5e1853a800534f76b90bc55df392/util/token.c) util which does random bytes, and [it's always used for the identifier no matter what](https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/c14aa1d0b82b5e1853a800534f76b90bc55df392/types/wlr_ext_foreign_toplevel_list_v1.c#L158-169). So, Sway ([usage](https://github.com/swaywm/sway/blob/f57c82a6f7cbd0c6499def867fc98459fb72a1d5/sway/tree/view.c#L822-L823)) and labwc ([usage](https://github.com/labwc/labwc/blob/359dd8f3c7a400a89613d013e5b9ba5647cdd476/src/foreign-toplevel/ext-foreign.c#L86-L87)) both do random bytes.

- Jay does something similar to wlroots: they have a [`ToplevelIdentifier`](https://github.com/mahkoh/jay/blob/5375dee0078b3060fb3fcd7cf8cab2430bd2fc84/src/utils/toplevel_identifier.rs#L14) newtype which is generated on object creation, and wraps a similar generic [`Opaque`](https://github.com/mahkoh/jay/blob/5375dee0078b3060fb3fcd7cf8cab2430bd2fc84/src/utils/opaque.rs#L19-L25) util.

- Hyprland exposes. uh. [the low bits of the client-specific `foreign_toplevel_list` object as well as the pointer to the window data](https://github.com/hyprwm/Hyprland/blob/bc764f7065c8e46b4e62cbb511b419034712c509/src/protocols/ForeignToplevel.cpp#L58). What? Isn't that just wrong? Yeah! It is. I tested it on my machine; two different clients on the same Hyprland will receive different identifiers for the same toplevel. It also doesn't list existing toplevels when a new list is created. Hyprland's implementation is just broken.

- Louvre has [a similar abstraction to that in this PR](https://github.com/CuarzoSoftware/Louvre/blob/f50fa742ad2cb87e64fe70a1f35c547c502b657c/src/lib/core/roles/LToplevelRole.h#L817-L827), and it works [exactly the same way](https://github.com/CuarzoSoftware/Louvre/blob/f50fa742ad2cb87e64fe70a1f35c547c502b657c/src/lib/core/roles/LToplevelRole.cpp#L121-L125): it's just a `u64` incremented for each mapped surface.

Also, [the initial implementation in COSMIC](https://github.com/i509VCB/cosmic-comp/blob/acba22544df7fac3bfe2b07cd9c0068d49da30d3/src/wayland/protocols/ext_foreign_toplevel_list.rs#L98) (which wasn't merged, but is the compositor linked on the protocol merge request) did the exact same random byte scheme as in present day COSMIC.

---

So, the opaque generational salt seems to be mostly a hypothetical. The main implementations sidestep it by relying on probability to guarantee two random identifiers aren't identical. Louvre *doesn't* address it. We already have the issue in `niri-ipc`. Nobody actually does a generation value. Niri has *absolutely zero* support for session handoff or anything else that would make this an issue in any other case than if the compositor crashes and some clients are left confused by that.

I reckon we just *don't* put a generation value. I think it's way nicer that the output of `lswt` and `niri msg windows` have the same readable identifier.

If we find in the future that such a generation value *needs to be included after all* (e.g. if we implement handoff?), then changing it does absolutely nothing to generic wayland clients (it's already opaque to them), and the niche case of any client parsing it and associating with `niri-ipc` will likely only affect a handful of clients who read our release notes, and the change on their end is fairly trivial.

YaLTeR commented:
Thanks for the overview of the identifiers. Guess let's hope nobody starts relying on the identifier being a number mapped ID like that.

> I've just noticed that [Hyprland has its own protocol to correlate foreign toplevel handles with IPC identifiers](https://wayland.app/protocols/hyprland-toplevel-mapping-v1). This is the only compositor that does so.
> 
> We could potentially offer a similar protocol. Is there any need to do so yet?

I don't think it's needed yet, no

YaLTeR commented on src/protocols/foreign_toplevel.rs:
I actually think that `Vec` with its `retain()` is faster than a `HashSet` for the small expected number of elements here (linear scan in a local memory region is very fast, and `HashSet` has to hash). But also it doesn't really matter and let's leave it as is; it's less code and it's clearer code this way.

YaLTeR reviewed (commented):
My remaining concern is the naming one that I mentioned above (`EXT_VERSION`-`WLR_VERSION`), other than that need to test it and then it's good
